acceptance/detector_test.go

```diff
@@ -0,0 +1,197 @@
+
+var (
+	detectDockerContext = filepath.Join("testdata", "detector")
+	detectorBinaryDir   = filepath.Join("testdata", "detector", "container", "cnb", "lifecycle")
+	detectImage         = "lifecycle/acceptance/detector"
+	userID              = "1234"
+)
+
+func TestDetector(t *testing.T) {
+	h.SkipIf(t, runtime.GOOS == "windows", "Detector is not yet supported on Windows")
+
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	h.MakeAndCopyLifecycle(t, "linux", detectorBinaryDir)
+	h.DockerBuild(t, detectImage, detectDockerContext)
+	defer h.DockerImageRemove(t, detectImage)
+
+	spec.Run(t, "acceptance-detector", testDetector, spec.Parallel(), spec.Report(report.Terminal{}))
+}
+
+func testDetector(t *testing.T, when spec.G, it spec.S) {
+	when("called with arguments", func() {
+		it("errors", func() {
+			cmd := exec.Command("docker", "run", "--rm", detectImage, "some-arg")
+			output, err := cmd.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to parse arguments: received unexpected arguments"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("running as a root", func() {
+		it("errors", func() {
+			cmd := exec.Command("docker", "run", "--rm", "--user", "root", detectImage)
+			output, err := cmd.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to build: refusing to run as root"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("read buildpack order file failed", func() {
+		it("errors", func() {
+			// no order.toml file in the default directory
+			cmd := exec.Command("docker", "run", "--rm", "--user", userID, detectImage)
+			output, err := cmd.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to read buildpack order file"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("no buildpack group passed detection", func() {
+		it("errors", func() {
+			cmd := exec.Command("docker", "run", "--rm", "--user", userID, "--env", "CNB_ORDER_PATH=/cnb/orders/empty_order.toml", detectImage)
+			output, err := cmd.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "No buildpack groups passed detection."
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	when("there is a buildpack group that pass detection", func() {
+		var copyDir, containerName string
+
+		it.Before(func() {
+			containerName = "test-container-" + h.RandString(10)
+			var err error
+			copyDir, err = ioutil.TempDir("", "test-docker-copy-")
+			h.AssertNil(t, err)
+		})
+
+		it.After(func() {
+			if h.DockerContainerExists(t, containerName) {
+				h.Run(t, exec.Command("docker", "rm", containerName))
+			}
+			os.RemoveAll(copyDir)
+		})
+
+		it("writes group.toml and plan.toml", func() {
+			h.DockerRunAndCopy(t,
+				containerName,
+				copyDir,
+				detectImage,
+				"/layers",
+				h.WithFlags("--user", userID,
+					"--env", "CNB_ORDER_PATH=/cnb/orders/simple_order.toml",
+				),
+				h.WithArgs(),
+			)
+
+			// check group.toml
```

Comment:
You could probably make a helper function (similar to what we use for analyzed.toml) to make this code terser and easier to follow.

Reply:
What would you put in this function? I think that there'll be too many arguments so the code won't be clearer.
What do you think?

Reply:
You could use `lifecycle.ReadGroup`. I know in general we don't want to use the code under test in the tests themselves but `lifecycle.ReadGroup` is a very simple helper.

Alternatively `toml.DecodeFile` could be used instead of `toml.Decode` to make this more terse.